drivers/sensor/lsm6dsr/lsm6dsr_trigger.c

```diff
@@ -0,0 +1,316 @@
+
+		lsm6dsr_read_reg(ctx, LSM6DSR_INT2_CTRL,
+				 (uint8_t *)&int2_ctrl, 1);
+		int2_ctrl.int2_drdy_g = enable;
+		return lsm6dsr_write_reg(ctx, LSM6DSR_INT2_CTRL,
+					 (uint8_t *)&int2_ctrl, 1);
+	}
+}
+
+/**
+ * lsm6dsr_trigger_set - link external trigger to event data ready
+ */
+int lsm6dsr_trigger_set(const struct device *dev,
+			  const struct sensor_trigger *trig,
+			  sensor_trigger_handler_t handler)
+{
+	const struct lsm6dsr_config *cfg = dev->config;
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+
+	__ASSERT_NO_MSG(trig->type == SENSOR_TRIG_DATA_READY);
+
+	if (!cfg->trig_enabled) {
+		LOG_ERR("trigger_set op not supported");
+		return -ENOTSUP;
+	}
+
+	if (trig->chan == SENSOR_CHAN_ACCEL_XYZ) {
+		lsm6dsr->handler_drdy_acc = handler;
+		lsm6dsr->trig_drdy_acc = trig;
+		if (handler) {
+			return lsm6dsr_enable_xl_int(dev, LSM6DSR_EN_BIT);
+		} else {
+			return lsm6dsr_enable_xl_int(dev, LSM6DSR_DIS_BIT);
+		}
+	} else if (trig->chan == SENSOR_CHAN_GYRO_XYZ) {
+		lsm6dsr->handler_drdy_gyr = handler;
+		lsm6dsr->trig_drdy_gyr = trig;
+		if (handler) {
+			return lsm6dsr_enable_g_int(dev, LSM6DSR_EN_BIT);
+		} else {
+			return lsm6dsr_enable_g_int(dev, LSM6DSR_DIS_BIT);
+		}
+	}
+#if defined(CONFIG_LSM6DSR_ENABLE_TEMP)
+	else if (trig->chan == SENSOR_CHAN_DIE_TEMP) {
+		lsm6dsr->handler_drdy_temp = handler;
+		lsm6dsr->trig_drdy_temp = trig;
+		if (handler) {
+			return lsm6dsr_enable_t_int(dev, LSM6DSR_EN_BIT);
+		} else {
+			return lsm6dsr_enable_t_int(dev, LSM6DSR_DIS_BIT);
+		}
+	}
+#endif
+
+	return -ENOTSUP;
+}
+
+/**
+ * lsm6dsr_handle_interrupt - handle the drdy event
+ * read data and call handler if registered any
+ */
+static void lsm6dsr_handle_interrupt(const struct device *dev)
+{
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+	const struct lsm6dsr_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lsm6dsr_status_reg_t status;
+
+	while (1) {
+		if (lsm6dsr_status_reg_get(ctx, &status) < 0) {
+			LOG_DBG("failed reading status reg");
+			return;
+		}
+
+		if ((status.xlda == 0) && (status.gda == 0)
+#if defined(CONFIG_LSM6DSR_ENABLE_TEMP)
+			&& (status.tda == 0)
+#endif
+		) {
+			break;
+		}
+
+		if ((status.xlda) && (lsm6dsr->handler_drdy_acc != NULL)) {
+			lsm6dsr->handler_drdy_acc(dev, lsm6dsr->trig_drdy_acc);
+		}
+
+		if ((status.gda) && (lsm6dsr->handler_drdy_gyr != NULL)) {
+			lsm6dsr->handler_drdy_gyr(dev, lsm6dsr->trig_drdy_gyr);
+		}
+
+#if defined(CONFIG_LSM6DSR_ENABLE_TEMP)
+		if ((status.tda) && (lsm6dsr->handler_drdy_temp != NULL)) {
+			lsm6dsr->handler_drdy_temp(dev, lsm6dsr->trig_drdy_temp);
+		}
+#endif
+	}
+
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy,
+					GPIO_INT_EDGE_TO_ACTIVE);
+}
+
+static inline void handle_irq(const struct device *dev)
+{
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+	const struct lsm6dsr_config *cfg = dev->config;
+
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy, GPIO_INT_DISABLE);
+
+#if defined(CONFIG_LSM6DSR_TRIGGER_OWN_THREAD)
```

Comment:
Please change the `#ifdef` to `if (IS_ENABLED(CONFIG_FOO)) {...}`

Reply:
I like this suggestion; I've changed that and in some other places as well! Thx

Reply:
You'll compilation errors if you try to protect code that uses conditionally defined struct members this way.  Also things like using global variables that don't exist, macros that aren't defined, functions that have differing prototypes.  While the code inside the if will be optimized away, it still needs to be valid.

Reply:
I did not noticed this because also the pipeline does not build all the cases, just the default one, but if own_thread is enabled, you can see the issue with this approach (by using `IS_ENABLED`):
```
/zephyr/drivers/sensor/lsm6dsr/lsm6dsr_trigger.c:226:39: error: 'struct lsm6dsr_data' has no member named 'work'
  226 |                 k_work_submit(&lsm6dsr->work);
      |                                       ^~
```

Reply:
Yeah, that's the tricky part - if there are configuration-dependent fields in the structure, this does not work.

The main reason I urge people to use the `IS_ENABLED()` approach, is that the configuration-dependent code is always at least compiled in all cases, even if it is discarded at a later stage. That makes a surprising difference when it comes to preventing regressions. There are some methods that are known to be good workarounds though - e.g. using a separate `struct` for `trigger` data and then having a pointer to it that is only used if `TRIGGER` is enabled.

Conditional compilation as well.

I'll see if I can make some additional suggestions.

Reply:
You'd still have the unused pointer wasting space in the struct.  Not that anyone cares about efficiency in STM sensor drivers anyway.  Might as well just remove all the configuration.

Reply:
It might be possible to use e.g. `COND_CODE_1()` to conditionally declare things.

Reply:
In any case, sorry for the delay reviewing this. The sensor area is not usually on my radar these days.